File: java/src/IceInternal/LocatorInfo.java
// **********************************************************************
//
// Copyright (c) 2003-2009 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package IceInternal;

public final class LocatorInfo
{
    private interface RequestCallback
    {
        public void response(LocatorInfo locatorInfo, Ice.ObjectPrx proxy);
        public void exception(LocatorInfo locatorInfo, Exception ex);
    };

    private abstract class Request
    {
        synchronized public void 
        addCallback(RequestCallback callback)
        {
            if(_response)
            {
                callback.response(_locatorInfo, _proxy);
                return;
            }
            else if(_exception != null)
            {
                callback.exception(_locatorInfo, _exception);
                return;
            }

            _callbacks.add(callback);

            if(!_sent)
            {
                _sent = true;
                try
                {
                    send();
                }
                catch(Exception ex)
                {
                    exception(ex);
                }
            }
        }

        synchronized public Ice.ObjectPrx 
        getProxy()
            throws Exception
        {
            if(_response)
            {
                return _proxy;
            }
            else if(_exception != null)
            {
                throw _exception;
            }

            if(!_sent)
            {
                _sent = true;
                try
                {
                    send();
                }
                catch(Exception ex)
                {
                    exception(ex);
                }
            }
    
            while(!_response && _exception == null)
            {
                try
                {
                    wait();
                }
                catch(java.lang.InterruptedException ex)
                {
                }
            }
    
            if(_exception != null)
            {
                throw _exception;
            }
            assert(_response);
            return _proxy;
        }

        protected 
        Request(LocatorInfo locatorInfo)
        {
            _locatorInfo = locatorInfo;
            _sent = false;
            _response = false;
        }

        synchronized protected void 
        response(Ice.ObjectPrx proxy)
        {
            _response = true;
            _proxy = proxy;
            for(RequestCallback c : _callbacks)
            {
                c.response(_locatorInfo, proxy);
            }
            notifyAll();
        }

        synchronized protected void 
        exception(Exception ex)
        {
            _exception = ex;
            for(RequestCallback c : _callbacks)
            {
                c.exception(_locatorInfo, ex);
            }
            notifyAll();
        }

        protected abstract void send();

        final protected LocatorInfo _locatorInfo;

        private java.util.List<RequestCallback> _callbacks = new java.util.ArrayList<RequestCallback>();
        private boolean _sent;
        private boolean _response;
        private Ice.ObjectPrx _proxy;
        private Exception _exception;
    };

    interface GetEndpointsCallback
    {
        void setEndpoints(EndpointI[] endpoints, boolean cached);
        void setException(Ice.LocalException ex);
    }

    private class ObjectRequest extends Request
    {
        public ObjectRequest(LocatorInfo locatorInfo, Ice.Identity id)
        {
            super(locatorInfo);
            _id = id;
        }

        protected void 
        send()
        {
            _locatorInfo.getLocator().findObjectById_async(
                new Ice.AMI_Locator_findObjectById()
                {
                    public void
                    ice_response(Ice.ObjectPrx proxy)
                    {
                        _locatorInfo.removeObjectRequest(_id);
                        response(proxy);
                    }

                    public void
                    ice_exception(Ice.UserException ex)
                    {
                        _locatorInfo.removeObjectRequest(_id);
                        exception(ex);
                    }
                    
                    public void
                    ice_exception(Ice.LocalException ex)
                    {
                        _locatorInfo.removeObjectRequest(_id);
                        exception(ex);
                    }
                },
                _id);
        }

        private final Ice.Identity _id;
    };

    private class AdapterRequest extends Request
    {
        public AdapterRequest(LocatorInfo locatorInfo, String id)
        {
            super(locatorInfo);
            _id = id;
        }
    
        protected void
        send()
        {
            _locatorInfo.getLocator().findAdapterById_async(
                new Ice.AMI_Locator_findAdapterById()
                {
                    public void
                    ice_response(Ice.ObjectPrx proxy)
                    {
                        _locatorInfo.removeAdapterRequest(_id);
                        response(proxy);
                    }

                    public void
                    ice_exception(Ice.UserException ex)
                    {
                        _locatorInfo.removeAdapterRequest(_id);
                        exception(ex);
                    }
                    
                    public void
                    ice_exception(Ice.LocalException ex)
                    {
                        _locatorInfo.removeAdapterRequest(_id);
                        exception(ex);
                    }
                },
                _id);
        }
        
        private final String _id;
    };

    private class ObjectRequestCallback implements RequestCallback
    {
        public void
        response(LocatorInfo locatorInfo, Ice.ObjectPrx object)
        {
            locatorInfo.getWellKnownObjectEndpoints(_reference, object, _ttl, false, _callback);
        }
        
        public void
        exception(LocatorInfo locatorInfo, Exception ex)
        {
            if(ex instanceof Ice.CollocationOptimizationException)
            {
                try
                {
                    Ice.BooleanHolder cached = new Ice.BooleanHolder();
                    _callback.setEndpoints(locatorInfo.getEndpoints(_reference, _ttl, cached), cached.value);
                }
                catch(Ice.LocalException e)
                {
                    _callback.setException(e);
                }
            }
            else
            {
                locatorInfo.getEndpointsException(_reference, ex, _callback);
            }
        }
    
        public 
        ObjectRequestCallback(Reference ref, int ttl, GetEndpointsCallback cb)
        {
            _reference = ref;
            _ttl = ttl;
            _callback = cb;
        }
    
        private final Reference _reference;
        private final int _ttl;
        private final GetEndpointsCallback _callback;
    };

    private class AdapterRequestCallback implements RequestCallback
    {
        public void
        response(LocatorInfo locatorInfo, Ice.ObjectPrx object)
        {
            EndpointI[] endpoints = null;
            if(object != null)
            {
                endpoints = ((Ice.ObjectPrxHelperBase)object).__reference().getEndpoints();
                if(endpoints.length > 0)
                {
                    locatorInfo.getTable().addAdapterEndpoints(_reference.getAdapterId(), endpoints);
                }
            }
            
            if(_reference.getInstance().traceLevels().location >= 1)
            {
                locatorInfo.getEndpointsTrace(_reference, endpoints, false);
            }
            
            if(endpoints == null)
            {
                _callback.setEndpoints(new EndpointI[0], false);
            }
            else
            {
                _callback.setEndpoints(endpoints, false);
            }
        }
        
        public void
        exception(LocatorInfo locatorInfo, Exception ex)
        {
            if(ex instanceof Ice.CollocationOptimizationException)
            {
                try
                {
                    Ice.BooleanHolder cached = new Ice.BooleanHolder();
                    _callback.setEndpoints(getEndpoints(_reference, _ttl, cached), cached.value);
                }
                catch(Ice.LocalException e)
                {
                    _callback.setException(e);
                }
            }
            else
            {
                getEndpointsException(_reference, ex, _callback);
            }
        }

        public 
        AdapterRequestCallback(Reference ref, int ttl, GetEndpointsCallback callback)
        {
            _reference = ref;
            _ttl = ttl;
            _callback = callback;
        }

        private final Reference _reference;
        private final int _ttl;
        private final GetEndpointsCallback _callback;
    };

    LocatorInfo(Ice.LocatorPrx locator, LocatorTable table)
    {
        _locator = locator;
        _table = table;
    }

    synchronized public void
    destroy()
    {
        _locatorRegistry = null;
        _table.clear();
    }

    public boolean
    equals(java.lang.Object obj)
    {
        if(this == obj)
        {
            return true;
        }

        if(obj instanceof LocatorInfo)
        {
            return _locator.equals(((LocatorInfo)obj)._locator);
        }

        return false;
    }

    public Ice.LocatorPrx
    getLocator()
    {
        //
        // No synchronization necessary, _locator is immutable.
        //
        return _locator;
    }

    public synchronized Ice.LocatorRegistryPrx
    getLocatorRegistry()
    {
        if(_locatorRegistry == null) // Lazy initialization
        {
            _locatorRegistry = _locator.getRegistry();

            //
            // The locator registry can't be located.
            //
            _locatorRegistry = Ice.LocatorRegistryPrxHelper.uncheckedCast(_locatorRegistry.ice_locator(null));
        }
        
        return _locatorRegistry;
    }

    public EndpointI[]
    getEndpoints(Reference ref, int ttl, Ice.BooleanHolder cached)
    {
        assert(ref.isIndirect());

        EndpointI[] endpoints = null;
        Ice.ObjectPrx object = null;
        cached.value = true;
        String adapterId = ref.getAdapterId();
        Ice.Identity identity = ref.getIdentity();
        
        try
        {
            if(adapterId.length() > 0)
            {
                endpoints = _table.getAdapterEndpoints(adapterId, ttl);
                if(endpoints == null)
                {
                    cached.value = false;
                    
                    if(ref.getInstance().traceLevels().location >= 1)
                    {
                        StringBuffer s = new StringBuffer();
                        s.append("searching for adapter by id\n");
                        s.append("adapter = " + adapterId);
                        ref.getInstance().initializationData().logger.trace(
                            ref.getInstance().traceLevels().locationCat, s.toString());
                    }

                    //
                    // Search the adapter in the location service if we didn't
                    // find it in the cache.
                    //
                    Request request = getAdapterRequest(adapterId);
                    object = request.getProxy();
                    if(object != null)
                    {
                        endpoints = ((Ice.ObjectPrxHelperBase)object).__reference().getEndpoints();
                        
                        if(endpoints.length > 0)
                        {
                            _table.addAdapterEndpoints(adapterId, endpoints);
                        }
                    }
                }
            }
            else
            {
                boolean objectCached = true;
                object = _table.getProxy(identity, ttl);
                if(object == null)
                {
                    objectCached = false;

                    if(ref.getInstance().traceLevels().location >= 1)
                    {
                        StringBuffer s = new StringBuffer();
                        s.append("searching for object by id\n");
                        s.append("object = " + ref.getInstance().identityToString(identity));
                        ref.getInstance().initializationData().logger.trace(
                            ref.getInstance().traceLevels().locationCat, s.toString());
                    }

                    Request request = getObjectRequest(identity);
                    object = request.getProxy();
                }
                
                boolean endpointsCached = true;
                if(object != null)
                {
                    Reference r = ((Ice.ObjectPrxHelperBase)object).__reference();
                    if(!r.isIndirect())
                    {
                        endpointsCached = false;
                        endpoints = r.getEndpoints();
                    }
                    else if(!r.isWellKnown())
                    {
                        Ice.BooleanHolder c = new Ice.BooleanHolder();
                        endpoints = getEndpoints(r, ttl, c);
                        endpointsCached = c.value;
                    }
                }
                
                if(!objectCached && endpoints != null && endpoints.length > 0)
                {
                    _table.addProxy(identity, object);
                }

                cached.value = objectCached || endpointsCached;
            }
        }
        catch(Exception ex)
        {
            getEndpointsException(ref, ex);
        }

        if(ref.getInstance().traceLevels().location >= 1)
        {
            getEndpointsTrace(ref, endpoints, cached.value);
        }

        return endpoints == null ? new EndpointI[0] : endpoints;
    }

    public void
    getEndpoints(final Reference ref, final int ttl, final GetEndpointsCallback callback)
    {
        final String adapterId = ref.getAdapterId();
        final Ice.Identity identity = ref.getIdentity();
        final Instance instance = ref.getInstance();
        if(adapterId.length() > 0)
        {
            EndpointI[] endpoints = _table.getAdapterEndpoints(adapterId, ttl);
            if(endpoints == null)
            {
                if(instance.traceLevels().location >= 1)
                {
                    StringBuffer s = new StringBuffer();
                    s.append("searching for adapter by id\n");
                    s.append("adapter = " + adapterId);
                    instance.initializationData().logger.trace(instance.traceLevels().locationCat, s.toString());
                }

                //
                // Search the adapter in the location service if we didn't
                // find it in the cache.
                //
                Request request = getAdapterRequest(adapterId);
                request.addCallback(new AdapterRequestCallback(ref, ttl, callback));
                return;
            }
            else
            {
                if(instance.traceLevels().location >= 1)
                {
                    getEndpointsTrace(ref, endpoints, true);
                }
                callback.setEndpoints(endpoints, true);
                return;
            }
        }
        else
        {
            Ice.ObjectPrx object = _table.getProxy(identity, ttl);
            if(object == null)
            {
                if(instance.traceLevels().location >= 1)
                {
                    StringBuffer s = new StringBuffer();
                    s.append("searching for object by id\n");
                    s.append("object = " + instance.identityToString(identity));
                    instance.initializationData().logger.trace(instance.traceLevels().locationCat, s.toString());
                }

                Request request = getObjectRequest(identity);
                request.addCallback(new ObjectRequestCallback(ref, ttl, callback));
                return;
            }
            else
            {
                getWellKnownObjectEndpoints(ref, object, ttl, true, callback);
                return;
            }
        }
    }

    public void
    clearObjectCache(Reference ref)
    {
        assert(ref.isIndirect());
        if(ref.isWellKnown())
        {
            Ice.ObjectPrx object = _table.removeProxy(ref.getIdentity());
            if(object != null)
            {
                Reference r = ((Ice.ObjectPrxHelperBase)object).__reference();
                if(!r.isIndirect())
                {
                    if(ref.getInstance().traceLevels().location >= 2)
                    {
                        trace("removed endpoints from locator table", ref, r.getEndpoints());
                    }
                }
                else if(!r.isWellKnown())
                {
                    clearCache(r);
                }
            }
        }
    }
    
    public void
    clearCache(Reference ref)
    {
        assert(ref.isIndirect());

        if(!ref.isWellKnown())
        {
            EndpointI[] endpoints = _table.removeAdapterEndpoints(ref.getAdapterId());

            if(endpoints != null && ref.getInstance().traceLevels().location >= 2)
            {
                trace("removed endpoints from locator table\n", ref, endpoints);
            }
        }
        else
        {
            Ice.ObjectPrx object = _table.removeProxy(ref.getIdentity());
            if(object != null)
            {
                Reference r = ((Ice.ObjectPrxHelperBase)object).__reference();
                if(!r.isIndirect())
                {
                    if(ref.getInstance().traceLevels().location >= 2)
                    {
                        trace("removed endpoints from locator table", ref, r.getEndpoints());
                    }
                }
                else if(!r.isWellKnown())
                {
                    clearCache(r);
                }
            }
        }
    }

    private void
    trace(String msg, Reference ref, EndpointI[] endpoints)
    {
        assert(ref.isIndirect());

        StringBuffer s = new StringBuffer();
        s.append(msg + "\n");
        if(!ref.isWellKnown())
        {
            s.append("adapter = " + ref.getAdapterId() + "\n");
        }
        else
        {
            s.append("object = " + ref.getInstance().identityToString(ref.getIdentity()) + "\n");
        }

        s.append("endpoints = ");
        final int sz = endpoints.length;
        for(int i = 0; i < sz; i++)
        {
            s.append(endpoints[i].toString());
            if(i + 1 < sz)
            {
                s.append(":");
            }
        }

        ref.getInstance().initializationData().logger.trace(ref.getInstance().traceLevels().locationCat, s.toString());
    }

    private void
    getEndpointsException(Reference ref, Exception exc)
    {
        assert(ref.isIndirect());

        try
        {
            throw exc;
        }
        catch(Ice.AdapterNotFoundException ex)
        {
            final Instance instance = ref.getInstance();
            if(instance.traceLevels().location >= 1)
            {
                StringBuffer s = new StringBuffer();
                s.append("adapter not found\n");
                s.append("adapter = " + ref.getAdapterId());
                instance.initializationData().logger.trace(instance.traceLevels().locationCat, s.toString());
            }

            Ice.NotRegisteredException e = new Ice.NotRegisteredException();
            e.kindOfObject = "object adapter";
            e.id = ref.getAdapterId();
            throw e;
        }
        catch(Ice.ObjectNotFoundException ex)
        {
            final Instance instance = ref.getInstance();
            if(instance.traceLevels().location >= 1)
            {
                StringBuffer s = new StringBuffer();
                s.append("object not found\n");
                s.append("object = " + instance.identityToString(ref.getIdentity()));
                instance.initializationData().logger.trace(instance.traceLevels().locationCat, s.toString());
            }

            Ice.NotRegisteredException e = new Ice.NotRegisteredException();
            e.kindOfObject = "object";
            e.id = instance.identityToString(ref.getIdentity());
            throw e;
        }
        catch(Ice.NotRegisteredException ex)
        {
            throw ex;
        }
        catch(Ice.LocalException ex)
        {
            final Instance instance = ref.getInstance();
            if(instance.traceLevels().location >= 1)
            {
                StringBuffer s = new StringBuffer();
                s.append("couldn't contact the locator to retrieve adapter endpoints\n");
                if(ref.getAdapterId().length() > 0)
                {
                    s.append("adapter = " + ref.getAdapterId() + "\n");
                }
                else
                {
                    s.append("object = " + instance.identityToString(ref.getIdentity()) + "\n");
                }
                s.append("reason = " + ex);
                instance.initializationData().logger.trace(instance.traceLevels().locationCat, s.toString());
            }
            throw ex;
        }
        catch(Exception ex)
        {
            assert(false);
        }
    }

    private void
    getEndpointsException(Reference ref, Exception exc, GetEndpointsCallback callback)
    {
        try
        {
            getEndpointsException(ref, exc);
        }
        catch(Ice.LocalException ex)
        {
            callback.setException(ex);
        }
        catch(Exception ex)
        {
            assert(false);
        }
    }

    private void
    getWellKnownObjectEndpoints(final Reference ref, 
                                final Ice.ObjectPrx object, 
                                final int ttl, 
                                final boolean objectCached, 
                                final GetEndpointsCallback callback)
    {
        EndpointI[] endpoints = null;
        if(object != null)
        {
            Reference r = ((Ice.ObjectPrxHelperBase)object).__reference();
            if(!r.isIndirect())
            {
                endpoints = r.getEndpoints();
            }
            else if(!r.isWellKnown())
            {
                getEndpoints(r, ttl, new GetEndpointsCallback()
                {
                    public void
                    setEndpoints(EndpointI[] endpoints, boolean endpointsCached)
                    {
                        if(!objectCached && endpoints != null && endpoints.length > 0)
                        {
                            _table.addProxy(ref.getIdentity(), object);
                        }
                        
                        if(ref.getInstance().traceLevels().location >= 1)
                        {
                            getEndpointsTrace(ref, endpoints, objectCached || endpointsCached);
                        }
                        
                        callback.setEndpoints(endpoints, objectCached || endpointsCached);
                    }
                    
                    public void
                    setException(Ice.LocalException ex)
                    {
                        callback.setException(ex);
                    }
                    });
                return;
            }
        }
        
        if(!objectCached && endpoints != null && endpoints.length > 0)
        {
            _table.addProxy(ref.getIdentity(), object);
        }

        if(ref.getInstance().traceLevels().location >= 1)
        {
            getEndpointsTrace(ref, endpoints, objectCached);
        }

        if(endpoints == null)
        {
            callback.setEndpoints(new EndpointI[0], false);
        }
        else
        {
            callback.setEndpoints(endpoints, objectCached);
        }
    }
 
    private void
    getEndpointsTrace(Reference ref, EndpointI[] endpoints, boolean cached)
    {
        if(endpoints != null && endpoints.length > 0)
        {
            if(cached)
            {
                trace("found endpoints in locator table", ref, endpoints);
            }
            else
            {
                trace("retrieved endpoints from locator, adding to locator table", ref, endpoints);
            }
        }
        else
        {
            final Instance instance = ref.getInstance();
            StringBuffer s = new StringBuffer();
            s.append("no endpoints configured for ");
            if(ref.getAdapterId().length() > 0)
            {
                s.append("adapter\n");
                s.append("adapter = " + ref.getAdapterId() + "\n");
            }
            else
            {
                s.append("object\n");
                s.append("object = " + instance.identityToString(ref.getIdentity()) + "\n");
            }
            instance.initializationData().logger.trace(instance.traceLevels().locationCat, s.toString());
        }
    }

    synchronized private Request
    getAdapterRequest(String id)
    {
        Request request = _adapterRequests.get(id);
        if(request != null)
        {
            return request;
        }
        
        request = new AdapterRequest(this, id);
        _adapterRequests.put(id, request);
        return request;
    }

    synchronized private void
    removeAdapterRequest(String id)
    {
        assert(_adapterRequests.get(id) != null);
        _adapterRequests.remove(id);
    }

    synchronized private Request
    getObjectRequest(Ice.Identity id)
    {
        Request request = _objectRequests.get(id);
        if(request != null)
        {
            return request;
        }
        
        request = new ObjectRequest(this, id);
        _objectRequests.put(id, request);
        return request;
    }

    synchronized private void
    removeObjectRequest(Ice.Identity id)
    {
        assert(_objectRequests.get(id) != null);
        _objectRequests.remove(id);
    }

    LocatorTable getTable()
    {
        return _table;
    }

    private final Ice.LocatorPrx _locator;
    private Ice.LocatorRegistryPrx _locatorRegistry;
    private final LocatorTable _table;

    private java.util.Map<String, Request> _adapterRequests = new java.util.HashMap<String, Request>();
    private java.util.Map<Ice.Identity, Request> _objectRequests = new java.util.HashMap<Ice.Identity, Request>();
}

File: cppe/src/IceE/LocatorInfo.cpp
// **********************************************************************
//
// Copyright (c) 2003-2007 ZeroC, Inc. All rights reserved.
//
// This copy of Ice-E is licensed to you under the terms described in the
// ICEE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <IceE/Config.h>

#ifdef ICEE_HAS_LOCATOR

#include <IceE/LocatorInfo.h>
#include <IceE/Locator.h>
#include <IceE/LocalException.h>
#include <IceE/Instance.h>
#include <IceE/TraceLevels.h>
#include <IceE/LoggerUtil.h>
#include <IceE/Endpoint.h>
#include <IceE/Reference.h>
#include <IceE/Functional.h>
#include <iterator>

using namespace std;
using namespace Ice;
using namespace IceInternal;

IceUtil::Shared* IceInternal::upCast(LocatorManager* p) { return p; }
IceUtil::Shared* IceInternal::upCast(LocatorInfo* p) { return p; }
IceUtil::Shared* IceInternal::upCast(LocatorTable* p) { return p; }

IceInternal::LocatorManager::LocatorManager() :
    _tableHint(_table.end())
{
}

void
IceInternal::LocatorManager::destroy()
{
    IceUtil::Mutex::Lock sync(*this);

    for_each(_table.begin(), _table.end(), Ice::secondVoidMemFun<const LocatorPrx, LocatorInfo>(&LocatorInfo::destroy));

    _table.clear();
    _tableHint = _table.end();

    _locatorTables.clear();
}

LocatorInfoPtr
IceInternal::LocatorManager::get(const LocatorPrx& loc)
{
    if(!loc)
    {
	return 0;
    }

    LocatorPrx locator = LocatorPrx::uncheckedCast(loc->ice_locator(0)); // The locator can't be located.

    //
    // TODO: reap unused locator info objects?
    //

    IceUtil::Mutex::Lock sync(*this);

    map<LocatorPrx, LocatorInfoPtr>::iterator p = _table.end();
    
    if(_tableHint != _table.end())
    {
	if(_tableHint->first == locator)
	{
	    p = _tableHint;
	}
    }
    
    if(p == _table.end())
    {
	p = _table.find(locator);
    }

    if(p == _table.end())
    {
	//
	// Rely on locator identity for the adapter table. We want to
	// have only one table per locator (not one per locator
	// proxy).
	//
	map<Identity, LocatorTablePtr>::iterator t = _locatorTables.find(locator->ice_getIdentity());
	if(t == _locatorTables.end())
	{
	    t = _locatorTables.insert(_locatorTables.begin(),
				      pair<const Identity, LocatorTablePtr>(locator->ice_getIdentity(),
				      new LocatorTable()));
	}

	_tableHint = _table.insert(_tableHint, 
				pair<const LocatorPrx, LocatorInfoPtr>(locator, new LocatorInfo(locator, t->second)));
    }
    else
    {
	_tableHint = p;
    }

    return _tableHint->second;
}

IceInternal::LocatorTable::LocatorTable()
{
}

void
IceInternal::LocatorTable::clear()
{
     IceUtil::Mutex::Lock sync(*this);

     _adapterEndpointsMap.clear();
     _objectMap.clear();
}

bool
IceInternal::LocatorTable::getAdapterEndpoints(const string& adapter, vector<EndpointPtr>& endpoints) const
{
    IceUtil::Mutex::Lock sync(*this);
    
    map<string, vector<EndpointPtr> >::const_iterator p = _adapterEndpointsMap.find(adapter);
    
    if(p != _adapterEndpointsMap.end())
    {
	endpoints = p->second;
	return true;
    }
    else
    {
	return false;
    }
}

void
IceInternal::LocatorTable::addAdapterEndpoints(const string& adapter, const vector<EndpointPtr>& endpoints)
{
    IceUtil::Mutex::Lock sync(*this);
    
    _adapterEndpointsMap.insert(make_pair(adapter, endpoints));
}

vector<EndpointPtr>
IceInternal::LocatorTable::removeAdapterEndpoints(const string& adapter)
{
    IceUtil::Mutex::Lock sync(*this);
    
    map<string, vector<EndpointPtr> >::iterator p = _adapterEndpointsMap.find(adapter);
    if(p == _adapterEndpointsMap.end())
    {
	return vector<EndpointPtr>();
    }

    vector<EndpointPtr> endpoints = p->second;

    _adapterEndpointsMap.erase(p);
    
    return endpoints;
}

bool 
IceInternal::LocatorTable::getProxy(const Identity& id, ObjectPrx& proxy) const
{
    IceUtil::Mutex::Lock sync(*this);
    
    map<Identity, ObjectPrx>::const_iterator p = _objectMap.find(id);
    
    if(p != _objectMap.end())
    {
	proxy = p->second;
	return true;
    }
    else
    {
	return false;
    }
}

void 
IceInternal::LocatorTable::addProxy(const Identity& id, const ObjectPrx& proxy)
{
    IceUtil::Mutex::Lock sync(*this);
    _objectMap.insert(make_pair(id, proxy));
}

ObjectPrx 
IceInternal::LocatorTable::removeProxy(const Identity& id)
{
    IceUtil::Mutex::Lock sync(*this);
    
    map<Identity, ObjectPrx>::iterator p = _objectMap.find(id);
    if(p == _objectMap.end())
    {
	return 0;
    }

    ObjectPrx proxy = p->second;
    _objectMap.erase(p);
    return proxy;
}

IceInternal::LocatorInfo::LocatorInfo(const LocatorPrx& locator, const LocatorTablePtr& table) :
    _locator(locator),
    _table(table)
{
    assert(_locator);
    assert(_table);
}

void
IceInternal::LocatorInfo::destroy()
{
    IceUtil::Mutex::Lock sync(*this);

    _locatorRegistry = 0;
    _table->clear();
}

bool
IceInternal::LocatorInfo::operator==(const LocatorInfo& rhs) const
{
    return _locator == rhs._locator;
}

bool
IceInternal::LocatorInfo::operator!=(const LocatorInfo& rhs) const
{
    return _locator != rhs._locator;
}

bool
IceInternal::LocatorInfo::operator<(const LocatorInfo& rhs) const
{
    return _locator < rhs._locator;
}

LocatorPrx
IceInternal::LocatorInfo::getLocator() const
{
    //
    // No mutex lock necessary, _locator is immutable.
    //
    return _locator;
}

LocatorRegistryPrx
IceInternal::LocatorInfo::getLocatorRegistry()
{
    IceUtil::Mutex::Lock sync(*this);
    
    if(!_locatorRegistry) // Lazy initialization.
    {
	_locatorRegistry = _locator->getRegistry();

	//
	// The locator registry can't be located.
	//
	_locatorRegistry = LocatorRegistryPrx::uncheckedCast(_locatorRegistry->ice_locator(0));
    }
    
    return _locatorRegistry;
}

vector<EndpointPtr>
IceInternal::LocatorInfo::getEndpoints(const IndirectReferencePtr& ref, bool& cached)
{
    vector<EndpointPtr> endpoints;
    ObjectPrx object;
    cached = true;    

    try
    {
	if(!ref->getAdapterId().empty())
	{
	    if(!_table->getAdapterEndpoints(ref->getAdapterId(), endpoints))
	    {
		cached = false;

                if(ref->getInstance()->traceLevels()->location >= 1)
                {
                    Trace out(ref->getInstance()->initializationData().logger,
                              ref->getInstance()->traceLevels()->locationCat);
                    out << "searching for adapter by id" << "\n";
                    out << "adapter = " << ref->getAdapterId();
                }
	    
		object = _locator->findAdapterById(ref->getAdapterId());
		if(object)
		{
		    endpoints = object->__reference()->getEndpoints();
		    _table->addAdapterEndpoints(ref->getAdapterId(), endpoints);
		}
	    }
	}
	else
	{
	    bool objectCached = true;
	    if(!_table->getProxy(ref->getIdentity(), object))
	    {

                if(ref->getInstance()->traceLevels()->location >= 1)
                {
                    Trace out(ref->getInstance()->initializationData().logger,
                              ref->getInstance()->traceLevels()->locationCat);
                    out << "searching for object by id" << "\n";
                    out << "object = " << ref->getInstance()->identityToString(ref->getIdentity());
                }

		objectCached = false;
		object = _locator->findObjectById(ref->getIdentity());
	    }

	    bool endpointsCached = true;
	    if(object)
	    {
		DirectReferencePtr odr = DirectReferencePtr::dynamicCast(object->__reference());
		if(odr)
		{
		    endpointsCached = false;
		    endpoints = odr->getEndpoints();
		}
		else
		{
		    IndirectReferencePtr oir = IndirectReferencePtr::dynamicCast(object->__reference());
		    assert(oir);
		    if(!oir->getAdapterId().empty())
		    {
			endpoints = getEndpoints(oir, endpointsCached);
		    }
		}
	    }

	    if(!objectCached && !endpoints.empty())
	    {
		_table->addProxy(ref->getIdentity(), object);
	    }

	    cached = objectCached || endpointsCached;
	}
    }
    catch(const AdapterNotFoundException&)
    {
        if(ref->getInstance()->traceLevels()->location >= 1)
        {
            Trace out(ref->getInstance()->initializationData().logger,
                      ref->getInstance()->traceLevels()->locationCat);
            out << "adapter not found" << "\n";
            out << "adapter = " << ref->getAdapterId();
        }

	NotRegisteredException ex(__FILE__, __LINE__);
	ex.kindOfObject = "object adapter";
	ex.id = ref->getAdapterId();
	throw ex;
    }
    catch(const ObjectNotFoundException&)
    {
        if(ref->getInstance()->traceLevels()->location >= 1)
        {
            Trace out(ref->getInstance()->initializationData().logger,
                      ref->getInstance()->traceLevels()->locationCat);
            out << "object not found" << "\n";
            out << "object = " << ref->getInstance()->identityToString(ref->getIdentity());
        }

	NotRegisteredException ex(__FILE__, __LINE__);
	ex.kindOfObject = "object";
	ex.id = ref->getInstance()->identityToString(ref->getIdentity());
	throw ex;
    }
    catch(const NotRegisteredException&)
    {
	throw;
    }
    catch(const LocalException& ex)
    {
	if(ref->getInstance()->traceLevels()->location >= 1)
	{
	    Trace out(ref->getInstance()->initializationData().logger, ref->getInstance()->traceLevels()->locationCat);
	    out << "couldn't contact the locator to retrieve adapter endpoints\n";
	    if(!ref)
	    {
		out << "object = " << ref->getInstance()->identityToString(ref->getIdentity()) << "\n";
	    }
	    else
	    {
		out << "adapter = " << ref->getAdapterId() << "\n";
	    }
	    out << "reason = " << ex.toString();
	}
	throw;
    }

    if(ref->getInstance()->traceLevels()->location >= 1)
    {
        if(!endpoints.empty())
        {
            if(cached)
            {
                trace("found endpoints in locator table", ref, endpoints);
            }
            else
            {
                trace("retrieved endpoints from locator, adding to locator table", ref, endpoints);
            }
        }
        else
        {
            Trace out(ref->getInstance()->initializationData().logger, ref->getInstance()->traceLevels()->locationCat);
            out << "no endpoints configured for ";
            if(ref->getAdapterId().empty())
            {
                out << "object\n";
                out << "object = " << ref->getInstance()->identityToString(ref->getIdentity());
            }
            else
            {
                out << "adapter\n";
                out << "adapter = " << ref->getAdapterId();
            }
        }
    }

    return endpoints;
}

void
IceInternal::LocatorInfo::clearObjectCache(const IndirectReferencePtr& ref)
{
    if(ref->getAdapterId().empty())
    {
	ObjectPrx object = _table->removeProxy(ref->getIdentity());
	if(object)
	{
	    IndirectReferencePtr oir = IndirectReferencePtr::dynamicCast(object->__reference());
	    if(oir)
	    {
	        if(!oir->getAdapterId().empty())
		{
		    clearCache(oir);
		}
	    }
	    else
	    {
		if(ref->getInstance()->traceLevels()->location >= 2)
		{
		    trace("removed endpoints from locator table", ref, object->__reference()->getEndpoints());
		}
	    }
	}
    }
}

void 
IceInternal::LocatorInfo::clearCache(const IndirectReferencePtr& ref)
{
    if(!ref->getAdapterId().empty())
    {
	vector<EndpointPtr> endpoints = _table->removeAdapterEndpoints(ref->getAdapterId());

	if(!endpoints.empty() && ref->getInstance()->traceLevels()->location >= 2)
	{
	    trace("removed endpoints from locator table", ref, endpoints);
	}
    }
    else
    {
	ObjectPrx object = _table->removeProxy(ref->getIdentity());
	if(object)
	{
	    IndirectReferencePtr oir = IndirectReferencePtr::dynamicCast(object->__reference());
	    if(oir)
	    {
	        if(!oir->getAdapterId().empty())
		{
		    clearCache(oir);
		}
	    }
	    else
	    {
		if(ref->getInstance()->traceLevels()->location >= 2)
		{
		    trace("removed endpoints from locator table", ref, object->__reference()->getEndpoints());
		}
	    }
	}
    }
}

void
IceInternal::LocatorInfo::trace(const string& msg,
	                        const IndirectReferencePtr& ref,
				const vector<EndpointPtr>& endpoints)
{
    Trace out(ref->getInstance()->initializationData().logger, ref->getInstance()->traceLevels()->locationCat);
    out << msg << "\n";
    if(!ref->getAdapterId().empty())
    {
	out << "adapter = "  << ref->getAdapterId() << "\n";
    }
    else
    {
	out << "object = "  << ref->getInstance()->identityToString(ref->getIdentity()) << "\n";
    }

    const char* sep = endpoints.size() > 1 ? ":" : "";
    out << "endpoints = ";
    for(unsigned int i = 0; i < endpoints.size(); ++i)
    {
        out << endpoints[i]->toString() << sep;
    }
}

#endif
